core/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java

```diff
@@ -0,0 +1,210 @@
+     *
+     * @param totalFrameWindowNanos total nanoseconds between when the first and last tasks were captured
+     * @param taskCount the total number of tasks that were executed
+     * @param totalTaskNanos total amount of time all tasks spent executing
+     */
+    static int calculateL(final long totalFrameWindowNanos, final int taskCount, final long totalTaskNanos) {
+        assert totalFrameWindowNanos > 0 : "cannot calculate for instantaneous tasks";
+        assert taskCount > 0 : "cannot calculate for no tasks";
+        // There is no set execution time, instead we adjust the time window based on the
+        // number of completed tasks, so there is no backwards thread required to update the
+        // queue size at a regular interval. This means we need to calculate our λ by the
+        // total runtime, rather than a fixed interval.
+
+        // λ = tasks per time period - total tasks divided by measurement time
+        final double lambda = (double) taskCount / totalFrameWindowNanos;
+
+        // W = avg task execution time - total task execution divided by total task count
+        final double w = totalTaskNanos / taskCount;
+
+        // L = λ * W
+        final int l = Math.toIntExact((long)(lambda * w));
+        return l;
+    }
+
+    /**
+     * Returns the current queue capacity
+     */
+    public int getCurrentCapacity() {
+        return workQueue.capacity();
+    }
+
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+        super.afterExecute(r, t);
+        // A task has been completed, it has left the building. We should now be able to get the
+        // total time as a combination of the time in the queue and time spent running the task. We
+        // only want runnables that did not throw errors though, because they could be fast-failures
+        // that throw off our timings, so only check when `t` is null.
+        if (t == null) {
+            assert r instanceof TimedRunnable : "expected only TimedRunnables in queue";
+            final long taskNanos = ((TimedRunnable) r).getTotalNanos();
+            final long totalNanos = totalTaskNanos.addAndGet(taskNanos);
+            if (taskCount.incrementAndGet() == this.tasksPerWindow) {
+                final long endTimeNs = System.nanoTime();
+                final long totalRuntime = endTimeNs - this.startNs;
+                // Reset the start time for all tasks. Theoretically this needs to be volatile, since
+                // we are reading from a different thread when it is set, but it is protected by the
+                // `taskCount` memory barrier.
+                startNs = endTimeNs;
+
+                // Calculate the new desired queue size
+                try {
+                    int desiredQueueSize = calculateL(totalRuntime, tasksPerWindow, totalNanos);
+                    logger.debug("estimated optimal queue size for [{}] threadpool: [{}]", name, desiredQueueSize);
```

Comment:
Nope, it generates logs like `estimated optimal queue size for [search] threadpool: [1000]`